File: game/internal/game.go
package internal

import (
	"encoding/json"
	"errors"
	"sync"

	"github.com/ahmetkoprulu/rtrp/game/models"
	"github.com/google/uuid"
)

type GameStatus string

const (
	GameStatusWaiting  GameStatus = "waiting"
	GameStatusStarting GameStatus = "starting"
	GameStatusStarted  GameStatus = "started"
	GameStatusEnding   GameStatus = "ending"
	GameStatusEnd      GameStatus = "end"
)

type GamePlayerStatus int

const (
	GamePlayerStatusWaiting GamePlayerStatus = iota
	GamePlayerStatusActive
	GamePlayerStatusInactive
	GamePlayerStatusFolded
)

type GameActionType string

const (
	GameActionTypePlayerJoin   GameActionType = "player_join"
	GameActionTypePlayerLeave  GameActionType = "player_leave"
	GameActionTypePlayerAction GameActionType = "player_action"
)

type GameError error

var (
	ErrorGameFull            GameError = errors.New("game_full")
	ErrorGamePlayerAlreadyIn GameError = errors.New("game_player_already_in")
	ErrorGamePositionTaken   GameError = errors.New("game_position_taken")
	ErrorGamePlayerNotFound  GameError = errors.New("game_player_not_found")
	ErrorGameNotReady        GameError = errors.New("game_not_ready")
)

type GameType int

const (
	GameTypeHoldem GameType = 1
)

type IPlayable interface {
	Start() error
	End() error
	OnPlayerJoin(player *GamePlayer) error
	OnPlayerLeave(player *GamePlayer) error
	ProcessAction(action json.RawMessage) error
	DealCards() error
	EvaluateHands() error
	CanStart() bool
	GetGameState() interface{}
}

type GameAction struct {
	PlayerID   string          `json:"player_id"`
	ActionType GameActionType  `json:"action_type"`
	Data       json.RawMessage `json:"data"`
}

type GameMessage struct {
	PlayerID    string          `json:"player_id"`
	ToGame      bool            `json:"to_game"`
	ToRoom      bool            `json:"to_room"`
	MessageType GameMessageType `json:"message_type"`
	Data        interface{}     `json:"data"`
}

type GameMessageType string

const (
	GameMessageTypePlayerAction GameMessageType = "player_action"
)

type IGamePlayer interface {
	GetBalance() int
	GetData() interface{}
}

type GamePlayer struct {
	Position   int              `json:"position"`
	Balance    int              `json:"balance"`
	LastAction string           `json:"last_action"`
	Client     *Client          `json:"client"`
	Status     GamePlayerStatus `json:"status"`
}

type Game struct {
	ID          string               `json:"id"`
	Status      GameStatus           `json:"status"`
	GameType    GameType             `json:"game_type"`
	Players     []*GamePlayer        `json:"players"`
	Playable    IPlayable            `json:"playable"`
	MinBet      int                  `json:"min_bet"`
	MaxPlayers  int                  `json:"max_players"`
	ActionChan  chan GameAction      `json:"-"`
	MessageChan chan models.Response `json:"-"`
	Room        *Room                `json:"-"`
	Mu          sync.RWMutex         `json:"-"`
}

func NewGame(actionChan chan GameAction, messageChan chan models.Response, room *Room, maxPlayers int, minBet int, gameType GameType) *Game {
	return &Game{
		ID:          uuid.New().String(),
		Status:      GameStatusWaiting,
		Players:     make([]*GamePlayer, 0),
		MaxPlayers:  maxPlayers,
		MinBet:      minBet,
		ActionChan:  actionChan,
		MessageChan: messageChan,
		Room:        room,
		GameType:    gameType,
		Mu:          sync.RWMutex{},
	}
}

func (g *Game) AddPlayer(position int, player *Client) error {
	g.Mu.Lock()
	defer g.Mu.Unlock()

	gamePlayer := &GamePlayer{
		Position: position,
		Client:   player,
		Balance:  int(player.User.Player.Chips),
	}

	if len(g.Players) >= g.MaxPlayers {
		return ErrorGameFull
	}

	for _, p := range g.Players {
		if p.Client.User.Player.ID == player.User.Player.ID {
			return ErrorGamePlayerAlreadyIn
		}

		if p.Position == position {
			return ErrorGamePositionTaken
		}
	}

	gamePlayer.Status = GamePlayerStatusWaiting
	g.Players = append(g.Players, gamePlayer)
	player.CurrentGame = g
	g.Playable.OnPlayerJoin(gamePlayer)

	return nil
}

func (g *Game) RemovePlayer(playerID string) error {
	g.Mu.Lock()
	defer g.Mu.Unlock()

	for _, p := range g.Players {
		if p.Client.User.Player.ID == playerID {
			// g.Players = slices.Delete(g.Players, i, i+1)
			p.Status = GamePlayerStatusInactive
			g.Playable.OnPlayerLeave(p)
			return nil
		}
	}
	return ErrorGamePlayerNotFound
}

func (g *Game) Start() error {
	if !g.Playable.CanStart() {
		return ErrorGameNotReady
	}

	err := g.Playable.Start()
	if err != nil {
		return err
	}

	g.Status = GameStatusStarted
	return nil
}

func (g *Game) End() error {
	err := g.Playable.End()
	if err != nil {
		return err
	}

	g.Status = GameStatusEnd
	return nil
}

func (g *Game) GetGameState() interface{} {
	return g.Playable.GetGameState()
}

File: Bot/WebSocket/Models/Game.cs
using Newtonsoft.Json;

namespace Websocket.Models;

public class Game
{
    [JsonProperty("id")] public string Id { get; set; }
    [JsonProperty("status")] public string Status { get; set; }
    [JsonProperty("gameType")] public string GameType { get; set; } // GameType
    [JsonProperty("players")] public List<GamePlayer> Players { get; set; } = new();
    [JsonProperty("minBet")] public int MinBet { get; set; }
    [JsonProperty("maxPlayers")] public int MaxPlayers { get; set; }
}

public class GamePlayer
{
    [JsonProperty("position")] public int Position { get; set; }
    [JsonProperty("balance")] public int Balance { get; set; }
    [JsonProperty("lastAction")] public string LastAction { get; set; }
    [JsonProperty("status")] public string Status { get; set; }
}

public class Card
{
    [JsonProperty("suit")] public string Suit { get; set; }
    [JsonProperty("value")] public string Value { get; set; }
    [JsonProperty("hidden")] public bool Hidden { get; set; }
}

public class GameMessage
{
    [JsonProperty("room_id")] public string RoomId { get; set; }
    [JsonProperty("player_id")] public string PlayerId { get; set; }
    [JsonProperty("game_type")] public GameType GameType { get; set; }
    [JsonProperty("data")] public object Data { get; set; }
}

public enum PlayerStatus
{
    Waiting,
    Active,
    Inactive,
}

public static class GameError
{
    public const string GameFull = "game_full";
    public const string GamePlayerAlreadyIn = "game_player_already_in";
    public const string GamePositionTaken = "game_position_taken";
    public const string GamePlayerNotFound = "game_player_not_found";
    public const string GameNotReady = "game_not_ready";
}

public static class GameStatus
{
    public const string Waiting = "waiting";
    public const string Starting = "starting";
    public const string Started = "started";
    public const string Ending = "ending";
    public const string End = "end";
}

public static class GameActionType
{
    public const string PlayerJoin = "player_join";
    public const string PlayerLeave = "player_leave";
    public const string PlayerAction = "player_action";
}

public enum GameType : short
{
    Holdem = 1,
}

public static class GameMessageType
{
    public const string GameStart = "game_start";
    public const string GameEnd = "game_end";
    public const string GamePlayerJoin = "game_player_join";
    public const string GamePlayerLeave = "game_player_leave";
}
